pasten.1
.\" Hey, EMACS: -*- nroff -*-
.TH PASTEN "1" "Nov 2013" "pasten"
.SH NAME
pasten \- Cassandra backed pastebin web application
.SH SYNOPSIS
.TP
.B pasten
\fI[\-\-auth\-server=host]\fR
\fI[\-\-bind=address]\fR
\fI[\-\-cacert=path]\fR
\fI[\-\-cassandra\-server=host]\fR
\fI[\-\-cert=path]\fR
\fI[\-\-corpus=name]\fR
\fI[\-\-doozer\-boot\-uri=URI]\fR
\fI[\-\-doozer\-uri=URI]\fR
\fI[\-\-exported\-name=name]\fR
\fI[\-\-help]\fR
\fI[\-\-key=path]\fR
\fI[\-\-keyspace=name]\fR
\fI[\-\-template\-dir=path]\fR
.SH DESCRIPTION
.PP
.B pasten
is a standalone web server implementing a pastebin with a
.I Cassandra
database backend.
.SH OPTIONS
.TP
\fB\-\-auth\-server=host\fR
Use the
.IR ancient\-login\-service (1)
service running on the given
.IR host .
The host may contain a port specification as required. As a protocol,
.I https
will always be used.
.TP
\fB\-\-bind=address\fR
Bind the server to the given address.
In
.IR doozer (1)
exported service mode, this should point to the externally reachable
host name or IP address of the host, otherwise this should be a host:port
pair.
.TP
\fB\-\-cacert=path\fR
Use the certificate file at the given path to verify peer certificates
(e.g. on the login token).
.TP
\fB\-\-cassandra\-server=host\fR
Use the Cassandra server at the given host (or host:port) for storage.
The server should have a keyspace with the name from
.I \-\-corpus
and be configured as described in the
.I cassandra-schema
file.
.TP
\fB\-\-cert=path\fR
.TP
\fB\-\-key=path\fR 
Use the certificate and key stored at the specified paths for signing
and encryption operations (such as the login cookie).
.TP
\fB\-\-corpus=name\fR
Use the Cassandra keyspace with the given name for storing pasted content.
.TP
\fB\-\-doozer\-boot\-uri=URI\fR
.TP
\fB\-\-doozer\-uri=URI\fR
Specifies how to connect to the
.IR doozer (1)
lock service, if used.
.IR doozer (1)
will be used only for the exported anonymous port functionality.
In order to use Doozer, you must specify a Doozer URI. 
The URI can contain an arbitrary number of address parts concatenated
by an ampersand character (&).
An URI always starts with the prefix
.IR doozer:? .
Valid options include:
.RS
.IP ca
specifies the "cluster address", which could be an IP address and port
combination or a host name with a port appended to it.
.IP cn
specifies a cluster name which will be looked up from the lock servers
specified in the environment variable
.B DOOZER_BOOT_URI
(which also contains a doozer URI).
.IP sk
can specify a shared key which is used to authenticate to the lock
service.
.RE
.RS
An example URI would look like this:
.TP
doozer:?ca=doozer1.example.com&ca=doozer2.example.com&sk=letmein
.RE
.TP
\fB\-\-exported\-name=name\fR
Export the host and port pair which
.B pasten
is bound to to a
.IR doozer (1)
lock service under the given service name.
.TP
\fB\-\-help\fR
Display the usage instructions for
.BR pasten .
.TP
\fB\-\-keyspace=name\fR
Set the keyspace to read from to
.IR name .
.TP
\fB\-\-template\-dir=path\fR
Indicate the path where the HTML template files for
.B pasten
can be found. Must point to a directory which contains the files
.IR display.tmpl ,
.IR error.tmpl ,
.I notfound.tmpl
and
.IR paste.tmpl .
.SH ENVIRONMENT
The following environment variables are honored:
.IP DOOZER_URI
The doozer cluster to bind to; overridden by
.IR --doozer-uri .
.IP DOOZER_BOOT_URI
The DzNS to lookup address in; overridden by
.IR --doozer-boot-uri .
.SH AUTHOR
.B pasten
was written by Tonnerre Lombard.
.SH "REPORTING BUGS"
Please file any bug in the ditz issue tracker contained in the
.B pasten
repository at
.I http://git.ancient-solutions.com/golang/pasten.git
.SH COPYRIGHT
.PP
Copyright \(co 2013 Tonnerre Lombard
.PP
Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:
.PP
.IP \[bu] 2
Redistributions of source code must retain the above copyright notice, this
list of conditions and the following disclaimer.
.IP \[bu] 2
Redistributions in binary form must reproduce the above copyright notice, this
list of conditions and the following disclaimer in the documentation and/or
other materials provided with the distribution.
.IP \[bu] 2
Neither the name of Ancient Solutions nor the name of its contributors may be
used to endorse or promote products derived from this software without specific
prior written permission.
.PP
THIS SOFTWARE IS PROVIDED BY ANCIENT SOLUTIONS AND CONTRIBUTORS ``AS IS'' AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS BE LIABLE FOR ANY
DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


ChangeLog
== 0.9 / 2013-12-22
* Fetch the entire row at a time in LookupPaste
* Use mutation pools in AddPaste
* Create README.md and manpage


pasten.go
/*
 * (c) 2013, Caoimhe Chaos <caoimhechaos@protonmail.com>,
 *	     Ancient Solutions. All rights reserved.
 *
 * Redistribution and use in source  and binary forms, with or without
 * modification, are permitted  provided that the following conditions
 * are met:
 *
 * * Redistributions of  source code  must retain the  above copyright
 *   notice, this list of conditions and the following disclaimer.
 * * Redistributions in binary form must reproduce the above copyright
 *   notice, this  list of conditions and the  following disclaimer in
 *   the  documentation  and/or  other  materials  provided  with  the
 *   distribution.
 * * Neither  the  name  of  Ancient Solutions  nor  the  name  of its
 *   contributors may  be used to endorse or  promote products derived
 *   from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS"  AND ANY EXPRESS  OR IMPLIED WARRANTIES  OF MERCHANTABILITY
 * AND FITNESS  FOR A PARTICULAR  PURPOSE ARE DISCLAIMED. IN  NO EVENT
 * SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL,  EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED  TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE,  DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT  LIABILITY,  OR  TORT  (INCLUDING NEGLIGENCE  OR  OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package main

import (
	"expvar"
	"flag"
	"log"
	"net/http"
	"os"
	"strings"
	"text/template"

	"ancient-solutions.com/ancientauth"
	"ancient-solutions.com/doozer/exportedservice"
)

var store *CassandraStore
var num_requests *expvar.Int = expvar.NewInt("num-requests")
var num_views *expvar.Int = expvar.NewInt("num-views")
var num_edits *expvar.Int = expvar.NewInt("num-edits")
var num_notfounds *expvar.Int = expvar.NewInt("num-notfounds")

var paste_templ *template.Template
var display_templ *template.Template
var error_templ *template.Template
var fourohfour_templ *template.Template
var authenticator *ancientauth.Authenticator

func Pasten(w http.ResponseWriter, req *http.Request) {
	var pasteid string = strings.Split(req.URL.Path, "/")[1]
	var err error

	num_requests.Add(1)

	if pasteid == "" {
		/* People need to be logged in in order to add URLs. */
		var user string = authenticator.GetAuthenticatedUser(req)
		var paste Paste

		// TODO(caoimhe): Count errors properly here.
		if user == "" {
			authenticator.RequestAuthorization(w, req)
			return
		}

		paste.User = user
		paste.Data = req.FormValue("paste")
		paste.Syntax = req.FormValue("syntax")
		paste.Title = req.FormValue("title")

		err = req.ParseForm()
		if err != nil {
			error_templ.Execute(w, err.Error())
			return
		}

		if paste.Data != "" {
			paste.Id, err = store.AddPaste(&paste, user)
			if err != nil {
				err = error_templ.Execute(w, err.Error())
				if err != nil {
					log.Print("Error executing error template: ", err)
				}
				return
			}
			num_edits.Add(1)
			err = display_templ.Execute(w, paste)
			if err != nil {
				log.Print("Error executing display template: ", err)
			}
			return
		}

		err = paste_templ.Execute(w, paste)
		if err != nil {
			log.Print("Error executing display template: ", err)
		}
	} else {
		var paste *Paste
		paste, err = store.LookupPaste(pasteid)
		if err != nil {
			err = error_templ.Execute(w, err.Error())
			if err != nil {
				log.Print("Error executing error template: ", err)
			}
			return
		}

		if paste == nil {
			w.WriteHeader(http.StatusNotFound)
			err = fourohfour_templ.Execute(w, pasteid)
			if err != nil {
				log.Print("Error executing 404 template: ", err)
			}
			num_notfounds.Add(1)
		} else {
			err = display_templ.Execute(w, paste)
			if err != nil {
				log.Print("Error executing display template: ", err)
			}
			num_views.Add(1)
		}
	}
}

func main() {
	var help bool
	var cassandra_server, keyspace, corpus string
	var ca, pub, priv, authserver string
	var bindto, templatedir, servicename string
	var doozer_uri, doozer_buri string
	var exporter *exportedservice.ServiceExporter
	var err error

	flag.BoolVar(&help, "help", false, "Display help")
	flag.StringVar(&bindto, "bind", "[::]:80",
		"The address to bind the web server to")
	flag.StringVar(&cassandra_server, "cassandra-server", "localhost:9160",
		"The Cassandra database server to use")
	flag.StringVar(&keyspace, "keyspace", "pasten",
		"The Cassandra keyspace the links are stored in. "+
			"The default should be fine.")
	flag.StringVar(&corpus, "corpus", "pastes",
		"The column family containing the paste data for this service")
	flag.StringVar(&ca, "cacert", "cacert.pem",
		"Path to the X.509 certificate of the certificate authority")
	flag.StringVar(&pub, "cert", "pasten.pem",
		"Path to the X.509 certificate")
	flag.StringVar(&priv, "key", "pasten.key",
		"Path to the X.509 private key file")
	flag.StringVar(&templatedir, "template-dir", "/var/www/templates",
		"Path to the HTML templates for the web interface")
	flag.StringVar(&authserver, "auth-server",
		"login.ancient-solutions.com",
		"The server to send the user to")
	flag.StringVar(&doozer_uri, "doozer-uri", os.Getenv("DOOZER_URI"),
		"Doozer URI to connect to")
	flag.StringVar(&doozer_buri, "doozer-boot-uri",
		os.Getenv("DOOZER_BOOT_URI"),
		"Doozer Boot URI to find named clusters")
	flag.StringVar(&servicename, "exported-name", "",
		"Name to export the service as in Doozer")
	flag.Parse()

	if help {
		flag.Usage()
		os.Exit(1)
	}
	paste_templ = template.Must(template.ParseFiles(templatedir +
		"/paste.tmpl"))
	display_templ = template.Must(template.ParseFiles(templatedir +
		"/display.tmpl"))
	error_templ = template.Must(template.ParseFiles(templatedir +
		"/error.tmpl"))
	fourohfour_templ = template.Must(template.ParseFiles(templatedir +
		"/notfound.tmpl"))

	authenticator, err = ancientauth.NewAuthenticator("Paste Bin", pub,
		priv, ca, authserver)
	if err != nil {
		log.Fatal("NewAuthenticator: ", err)
	}

	store = NewCassandraStore(cassandra_server, keyspace, corpus)
	if store == nil {
		os.Exit(2)
	}

	http.Handle("/", http.HandlerFunc(Pasten))

	if len(servicename) > 0 {
		exporter, err = exportedservice.NewExporter(doozer_uri,
			doozer_buri)
		if err != nil {
			log.Fatal("NewExporter: ", err)
		}

		err = exporter.ListenAndServeNamedHTTP(servicename, bindto,
			nil)
		if err != nil {
			log.Fatal("ListenAndServeNamedHTTP: ", err)
		}
	} else {
		err = http.ListenAndServe(bindto, nil)
		if err != nil {
			log.Fatal("ListenAndServe: ", err)
		}
	}
}
